## Business Question

Can we construct a clean, analysis-ready dataset from raw transactional tables that fully reconciles with SQL outputs and can be reliably used for deeper analysis?

---

## Context

The SQL phase established:
- Discounting is the dominant pricing behavior (~61% of transactions)
- Margin compression exists as discount depth increases
- Patterns are stable across categories, customers, stores, and time

This notebook focuses on:
- Rebuilding the analytical dataset in Python
- Validating consistency with SQL results
- Ensuring data integrity before further analysis

---

## Scope

- Load data from PostgreSQL
- Join core tables into a unified dataset
- Recreate key business metrics
- Perform validation checks against SQL outputs

---

## Output

A clean, validated dataset ready for:
- Statistical analysis
- Segmentation
- Visualization

### Import Libraries

In [2]:
import pandas as pd
import numpy as np

from sqlalchemy import create_engine

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

### Data Source Connection

Connecting to PostgreSQL to extract required tables for analysis.

In [3]:
# Update credentials accordingly
engine = create_engine("postgresql://postgres:zeeshan786@localhost:5432/contoso_retail")

# Load required tables (excluding SALES)
orderrows = pd.read_sql("SELECT * FROM orderrows", engine)
orders = pd.read_sql("SELECT * FROM orders", engine)
customer = pd.read_sql("SELECT * FROM customer", engine)
product = pd.read_sql("SELECT * FROM product", engine)
store = pd.read_sql("SELECT * FROM store", engine)

### Initial Data Overview

Quick structural checks to confirm:
- Row counts
- Column presence
- Alignment with SQL phase

In [4]:
tables = {
    "orderrows": orderrows,
    "orders": orders,
    "customer": customer,
    "product": product,
    "store": store
}

for name, df in tables.items():
    print(f"\n{name.upper()}")
    print(f"Shape: {df.shape}")
    print(df.head(2))


ORDERROWS
Shape: (223974, 7)
   orderkey  linenumber  productkey  quantity  unitprice  netprice  unitcost
0    139000           0         153         2     375.98    375.98    172.90
1    139000           1        1621         8       7.79      7.79      3.97

ORDERS
Shape: (93470, 6)
   orderkey  customerkey  storekey   orderdate deliverydate currencycode
0    139000      1293529       430  2016-05-18   2016-05-18          USD
1    139001       915563       410  2016-05-18   2016-05-18          GBP

CUSTOMER
Shape: (104990, 24)
   customerkey  geoareakey     startdt       enddt  continent  gender title  \
0           15           4  1990-09-10  2034-07-29  Australia    male   Mr.   
1           23           8  1995-08-11  2045-01-26  Australia  female   Ms.   

  givenname middleinitial   surname  ... zipcode country countryfull  \
0    Julian             A  McGuigan  ...    4357      AU   Australia   
1      Rose             H      Dash  ...    6055      AU   Australia   

     birt

### Observations

- Row counts match SQL baseline across all tables
- Schema and data types align with expectations
- `store.status` contains nulls; `closedate` null for active stores (expected behavior)
- Multi-currency present; no normalization applied
- Customer table includes unused demographic fields

### Assumptions

- Primary keys remain unique (validated in SQL phase)
- Raw data reflects source truth (no transformations applied)
- Null `store.status` treated as active unless explicitly filtered
- `customer.age` not used due to potential staleness

### Limitations

- No DB-level constraints enforced; relies on prior SQL validation
- Store lifecycle fields require explicit null handling
- Currency not normalized; analysis limited to ratio-based metrics

### Join Strategy

The analytical dataset will be constructed incrementally to avoid row duplication and ensure join integrity.

Join path:

- `orderrows` -> base transactional grain (OrderKey, LineNumber)
- `orders` -> adds CustomerKey, StoreKey, OrderDate, CurrencyCode
- `product` -> adds category and product attributes
- `customer` -> adds customer attributes
- `store` -> adds store attributes

All joins are expected to be **many-to-one** relative to `orderrows`.

Key validation principle:
Row count must remain constant after each join (223,974).

### Base Join (orderrows + orders)

In [5]:
df = orderrows.merge(
    orders,
    on="orderkey",
    how="left",
    validate="many_to_one"
)

print("Row count after orders join:", df.shape[0])

Row count after orders join: 223974


### Validation -> Orders Join
- Row count remains 223,974, no duplication or drop
- `many_to_one` validation passed, confirming each order line maps to exactly one order
- Join integrity confirmed; proceeding to next merge

### Add Product

In [6]:
df = df.merge(
    product,
    on="productkey",
    how="left",
    validate="many_to_one"
)

print("Row count after product join:", df.shape[0])

Row count after product join: 223974


### Add Customer

In [7]:
df = df.merge(
    customer,
    on="customerkey",
    how="left",
    validate="many_to_one"
)

print("Row count after customer join:", df.shape[0])

Row count after customer join: 223974


### Add Store

In [8]:
df = df.merge(
    store,
    on="storekey",
    how="left",
    validate="many_to_one"
)

print("Final row count:", df.shape[0])

Final row count: 223974


### Validation Check

In [9]:
expected_rows = 223974

if df.shape[0] != expected_rows:
    print("Row count mismatch. STOP")
else:
    print("Row count preserved, joins are safe")

Row count preserved, joins are safe


### Quantitative null check

In [11]:
df.isnull().mean().sort_values(ascending=False).head(10)

status          0.991794
closedate       0.987833
squaremeters    0.417682
weight          0.149727
weightunit      0.059346
company         0.000888
productkey      0.000000
quantity        0.000000
unitprice       0.000000
netprice        0.000000
dtype: float64

### Null Check (Integrity Check)

In [10]:
df[[
    "customerkey",
    "productkey",
    "storekey",
    "orderdate"
]].isnull().sum()

customerkey    0
productkey     0
storekey       0
orderdate      0
dtype: int64

### Join Integrity Observations
- Row count preserved at 223,974 after all joins
- All merges validated as many-to-one; no duplication introduced
- No nulls in critical keys (customerkey, productkey, storekey, orderdate)
- `status` (99.2%) and `closedate` (98.8%) carry high null rates, expected, as both are store-level fields that apply only to closed stores, representing a small fraction of the dataset
- `squaremeters` (41.8%) and `weight` (15.0%) nulls are noted but do not affect core financial metrics
- `company` (0.09%) has marginal nulls in the customer dimension, acceptable for analysis purposes

### Assumptions
- Dimension tables fully map to transaction records
- No orphan records exist
- High null rates in store lifecycle fields reflect data design, not quality failure

### Limitations
- No database-enforced constraints; integrity verified via observed checks
- Columns with structural nulls (status, closedate) require explicit handling if store lifecycle filtering is applied

### Persist the dataset

In [25]:
df.to_pickle(r"D:\data-analysis-projects\retail-discount-profitability-audit\exports\fact_sales_enriched.pkl")

- Dataset persisted as pickle for interim use; not intended for cross-environment sharing

### Metric Reconstruction

Recreating core business metrics in Python to validate consistency with SQL outputs.

Metrics defined:

- Revenue = Quantity × NetPrice  
- Cost = Quantity × UnitCost  
- Profit = Revenue − Cost  
- Margin % = Profit / Revenue  
- Discount Flag = NetPrice < UnitPrice OR NetPrice != UnitPrice

These must reconcile exactly with SQL aggregates before proceeding.

### Create Metrics

In [28]:
df["revenue"] = df["quantity"] * df["netprice"]
df["cost"] = df["quantity"] * df["unitcost"]
df["profit"] = df["revenue"] - df["cost"]

df["margin_pct"] = (df["profit"] / df["revenue"]).round(2)

df["is_discounted"] = df["netprice"] < df["unitprice"]

### Aggregate Check

In [31]:
total_revenue = df["revenue"].sum()
total_cost = df["cost"].sum()
total_profit = df["profit"].sum()

discount_pct = df["is_discounted"].mean() * 100

print("Revenue:", round(total_revenue, 2))
print("Cost:", round(total_cost, 2))
print("Profit:", round(total_profit, 2))
print("Discount %:", round(discount_pct, 2))

Revenue: 218814485.15
Cost: 96482453.47
Profit: 122332031.68
Discount %: 61.13


### Validation - Metric Consistency
- Do Python aggregates match SQL results exactly (within rounding tolerance)?
- Any discrepancy indicates a join issue, calculation inconsistency, or data type precision problem

### Observations
- All four metrics match SQL baseline exactly
- Revenue, cost, and profit confirm join integrity and calculation consistency
- Discount rate of 61% confirms `is_discounted` flag logic aligns with SQL definition (`netprice < unitprice`)

### Assumptions
- Calculations mirror SQL logic exactly
- No transformation has altered underlying values

### Limitations
- Floating-point precision may introduce negligible rounding differences